File: src/main/java/com/oriaxx77/oaths/domain/service/UserService.scala
package com.oriaxx77.oaths.domain.service

import java.util.UUID
import scala.collection._
import org.springframework.stereotype.Component
import scala.collection.concurrent.TrieMap
import com.oriaxx77.oaths.domain.enity.User
import org.springframework.beans.factory.annotation.Autowired
import com.oriaxx77.oaths.domain.repository.UserRepository


// TODO: refactor. Looks like a transaction script :(
// TODO: extract interface
@Component
class UserService @Autowired() ( userRepository: UserRepository ) {
  
  var emailUserMap = new TrieMap[String,User];
  
  /** 
   *  Creates a user based on it's email.
   *  Returns with an authToken.
   */
  def create( email: String, userName: String ): String = {
    
    // TODO: use ScalaZ for validation
    // TODO: send auth email instead of returning the token.
    // TODO: token should not be created here.
    if ( !Option(email).isDefined || email.length() == 0 )
      throw new IllegalArgumentException( "Missing email!" )
    
    val userForEmail = userRepository.findUserByEmail( email )
    if ( userForEmail.isPresent() ){
      return userForEmail.get.authToken
    }
    
    val newUser = User( email,
                    userName,
                    UUID.randomUUID().toString(),
                    UUID.randomUUID().toString(),
                    "" )
    userRepository.save( newUser )
    // TODO: return whole User?
    return newUser.authToken
  }
  
  /** Returns authToken */
  def authByEmail( email: String ): String = {
    // TODO: use ScalaZ for validation
    // TODO: refactor to send an auth email instead.
    val user = userRepository.findUserByEmail( email )
    if ( user.isPresent() ) {
      throw new RuntimeException( "Not found" );
    }
    
    user.get.authToken =  UUID.randomUUID().toString()
    userRepository.save( user.get )
    return user.get.authToken 
  }
  
  
  /** Returns authToken */
  def authByCode( authCode: String ): String = {
    // TODO: use ScalaZ for validation
    val optUser = userRepository.findUserByAuthCode(authCode)
    if ( optUser.isPresent() ) {
      throw new RuntimeException( "Not found" );
    }
    val user = optUser.get
    user.authToken =  UUID.randomUUID().toString()
    userRepository.save( user )
    return user.authToken
  }
  
  
  /** Sets user.pushNotificationDeviceToken */
  def setPushNotificationDeviceToken( authToken: String,
                                      deviceToken: String ) { 
    // TODO:use ScalaZ for validation  
    // TODO: use ScalaZ for validation
    val optUser = userRepository.findUserByAuthToken( authToken )
    if ( optUser.isPresent() ) {
      throw new RuntimeException( "Not found" );
    }
    val user = optUser.get
    user.pushNotifictionDeviceToken = deviceToken
    userRepository.save( user )
  }
                          
  
  
}

File: src/main/java/com/oriaxx77/oaths/domain/service/UserService.java
package com.oriaxx77.oaths.domain.service;

import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;


import com.oriaxx77.oaths.domain.enity.User;
import com.oriaxx77.oaths.domain.repository.UserRepository;



// TODO: refactor. Looks like a transaction script :(
// TODO: extract interface
@Component
public class UserService {
  
	private UserRepository userRepository;
	
	@Autowired
	public UserService( UserRepository userRepository ){
		this.userRepository = userRepository;
	}
	
  
  /** 
   *  Creates a user based on it's email.
   *  Returns with an authToken.
   */
  public String create( String email, String userName) {
    
    // TODO: use ScalaZ for validation
    // TODO: send auth email instead of returning the token.
    // TODO: token should not be created here.
    if ( email == null || email.length() == 0 )
      throw new IllegalArgumentException( "Missing email!" );
    
    return  userRepository.findUserByEmail( email )
			  			  .orElseGet( () -> { 
			  				  User newUser = new User( email, userName,    			  			  
			  										   UUID.randomUUID().toString(),
			  										   UUID.randomUUID().toString(),"" );
			  				  userRepository.save( newUser ); // TODO: sideffect!!!!
			  				  return newUser;})
			  			  .getAuthToken();    
  }
  
  /** Returns authToken */
  public String authByEmail( String email ) {
    // TODO: use ScalaZ for validation
    // TODO: refactor to send an auth email instead.
    User user = userRepository.findUserByEmail( email )
    								.orElseThrow( () -> new RuntimeException( "Not found" ) );
    user.setAuthToken( UUID.randomUUID().toString() );
    userRepository.save( user );	// WHY? It is [User,Long]
    return user.getAuthToken(); 
  }
  
  
  /** Returns authToken */
  public String authByCode( String authCode){
    // TODO: use ScalaZ for validation
    User user = userRepository.findUserByAuthCode( authCode )
    										.orElseThrow( () -> new RuntimeException( "Not found" ) );    
    user.setAuthToken( UUID.randomUUID().toString() );
    userRepository.save( user );	// WHY? It is [User,Long]
	return user.getAuthToken(); 
  }
  
  
  /** Sets user.pushNotificationDeviceToken */
  public void setPushNotificationDeviceToken( String authToken,
                                      	 String deviceToken ) { 
    // TODO:use ScalaZ for validation  
    // TODO: use ScalaZ for validation
    User user = userRepository.findUserByAuthToken( authToken )
    						  .orElseThrow( () -> new RuntimeException("Not found") );
    user.setPushNotifictionDeviceToken( deviceToken );
    userRepository.save( user );
  }
                          
  
  
}